File: backend/app/schemas/jd_intelligence.py
from datetime import date, datetime
from enum import Enum
from typing import Dict, List, Optional

from pydantic import BaseModel, Field, HttpUrl


class JDSourceType(str, Enum):
    pasted_text = "pasted_text"
    link = "link"
    screenshot_ocr = "screenshot_ocr"
    recruiter_message = "recruiter_message"
    manual_note = "manual_note"
    imported_xlsx = "imported_xlsx"


class RoleOrientation(str, Enum):
    research = "research"
    engineering = "engineering"
    research_engineering = "research_engineering"
    tpm_variant = "tpm_variant"
    unclear = "unclear"


class TransitionAcceptance(str, Enum):
    yes = "yes"
    no = "no"
    depends = "depends"
    unclear = "unclear"


class ApplicationPriority(str, Enum):
    high = "high"
    medium = "medium"
    low = "low"
    not_apply = "not_apply"
    unknown = "unknown"


class JDEntryStatus(str, Enum):
    new = "new"
    watching = "watching"
    analyzed = "analyzed"
    favorite = "favorite"
    contacting = "contacting"
    paused = "paused"
    rejected = "rejected"
    archived = "archived"


class InterestLevel(str, Enum):
    dislike = "dislike"
    neutral = "neutral"
    interested = "interested"
    favorite = "favorite"


class FitFeeling(str, Enum):
    exciting = "exciting"
    practical = "practical"
    risky = "risky"
    unclear = "unclear"


class FollowUpIntent(str, Enum):
    no_action = "no_action"
    keep_watching = "keep_watching"
    ask_recruiter = "ask_recruiter"
    build_contact_only = "build_contact_only"
    prepare_later = "prepare_later"


class SkillCategory(str, Enum):
    localization = "localization"
    slam = "slam"
    reconstruction = "reconstruction"
    world_model = "world_model"
    deep_learning_training = "deep_learning_training"
    engineering = "engineering"
    research_writing = "research_writing"
    interview = "interview"


class CapabilityLevel(str, Enum):
    none = "none"
    beginner = "beginner"
    working = "working"
    solid = "solid"
    strong = "strong"
    expert = "expert"


class ConfidenceLevel(str, Enum):
    low = "low"
    medium = "medium"
    high = "high"


class PriorityLevel(str, Enum):
    high = "high"
    medium = "medium"
    low = "low"


class TaskCategory(str, Enum):
    reconstruction_line = "reconstruction_line"
    world_model_line = "world_model_line"
    direction_decision = "direction_decision"
    paper_writing = "paper_writing"
    resume_building = "resume_building"
    jd_follow_up = "jd_follow_up"
    habit = "habit"


class TaskStatus(str, Enum):
    planned = "planned"
    active = "active"
    blocked = "blocked"
    paused = "paused"
    completed = "completed"
    archived = "archived"


class Adjustability(str, Enum):
    low = "low"
    medium = "medium"
    high = "high"


class JDRoleType(str, Enum):
    research_engineer = "research_engineer"
    algorithm_engineer = "algorithm_engineer"
    research_engineering = "research_engineering"
    technical_lead = "technical_lead"
    tpm_variant = "tpm_variant"
    business_operation = "business_operation"
    unclear = "unclear"


class MatchLevel(str, Enum):
    high = "high"
    medium_high = "medium_high"
    medium = "medium"
    low = "low"
    not_match = "not_match"


class TimingRecommendation(str, Enum):
    apply_now = "apply_now"
    build_contact_only = "build_contact_only"
    pause_and_prepare = "pause_and_prepare"
    do_not_apply = "do_not_apply"


class RiskLevel(str, Enum):
    low = "low"
    medium = "medium"
    high = "high"
    unknown = "unknown"


class CandidateActionType(str, Enum):
    learn = "learn"
    read_paper = "read_paper"
    run_experiment = "run_experiment"
    update_resume = "update_resume"
    contact_recruiter = "contact_recruiter"
    collect_more_jds = "collect_more_jds"
    prepare_interview_answer = "prepare_interview_answer"
    update_tech_radar_scope = "update_tech_radar_scope"


class TimeSensitivity(str, Enum):
    this_week = "this_week"
    this_month = "this_month"
    later = "later"


class CandidateActionStatus(str, Enum):
    suggested = "suggested"
    accepted = "accepted"
    rejected = "rejected"
    deferred = "deferred"
    converted_to_task = "converted_to_task"
    done = "done"


class JDEntryBase(BaseModel):
    company: str = ""
    team_or_department: str = ""
    role_title: str = ""
    city: str = ""
    source_type: JDSourceType = JDSourceType.pasted_text
    source_name: str = ""
    source_url: Optional[HttpUrl] = None
    record_date: date
    jd_text: str = ""
    recruiter_context: str = ""
    notes: str = ""
    must_have_skills: List[str] = []
    bonus_skills: List[str] = []
    new_keywords: List[str] = []
    salary_range: str = ""
    salary_min_k: Optional[int] = None
    salary_max_k: Optional[int] = None
    salary_months: str = ""
    equity: str = ""
    education_requirement: str = ""
    major_requirement: str = ""
    experience_years: str = ""
    paper_or_patent_requirement: str = ""
    domain_experience_requirement: str = ""
    role_orientation: RoleOrientation = RoleOrientation.unclear
    accepts_transition: TransitionAcceptance = TransitionAcceptance.unclear
    language_or_overseas_requirement: str = ""
    month_change: str = ""
    personal_gap_assessment: str = ""
    application_priority: ApplicationPriority = ApplicationPriority.unknown
    status: JDEntryStatus = JDEntryStatus.new


class JDEntryCreate(JDEntryBase):
    record_date: date = Field(default_factory=date.today)


class JDImageImportRequest(BaseModel):
    image_base64: str
    filename: str = ""
    source_name: str = "image_upload"


class JDImageImportResult(BaseModel):
    ocr_text: str
    draft_entry: JDEntryCreate
    confidence: ConfidenceLevel = ConfidenceLevel.medium
    notes: str = ""


class JDEntryUpdate(BaseModel):
    company: Optional[str] = None
    team_or_department: Optional[str] = None
    role_title: Optional[str] = None
    city: Optional[str] = None
    source_type: Optional[JDSourceType] = None
    source_name: Optional[str] = None
    source_url: Optional[HttpUrl] = None
    record_date: Optional[date] = None
    jd_text: Optional[str] = None
    recruiter_context: Optional[str] = None
    notes: Optional[str] = None
    must_have_skills: Optional[List[str]] = None
    bonus_skills: Optional[List[str]] = None
    new_keywords: Optional[List[str]] = None
    salary_range: Optional[str] = None
    salary_min_k: Optional[int] = None
    salary_max_k: Optional[int] = None
    salary_months: Optional[str] = None
    equity: Optional[str] = None
    education_requirement: Optional[str] = None
    major_requirement: Optional[str] = None
    experience_years: Optional[str] = None
    paper_or_patent_requirement: Optional[str] = None
    domain_experience_requirement: Optional[str] = None
    role_orientation: Optional[RoleOrientation] = None
    accepts_transition: Optional[TransitionAcceptance] = None
    language_or_overseas_requirement: Optional[str] = None
    month_change: Optional[str] = None
    personal_gap_assessment: Optional[str] = None
    application_priority: Optional[ApplicationPriority] = None
    status: Optional[JDEntryStatus] = None


class JDEntry(JDEntryBase):
    id: str
    created_at: datetime
    updated_at: datetime


class JDPreferenceMarkCreate(BaseModel):
    interest_level: InterestLevel = InterestLevel.neutral
    fit_feeling: FitFeeling = FitFeeling.unclear
    user_tags: List[str] = []
    why_liked: str = ""
    why_hesitated: str = ""
    follow_up_intent: FollowUpIntent = FollowUpIntent.no_action


class JDPreferenceMark(JDPreferenceMarkCreate):
    id: str
    jd_entry_id: str
    created_at: datetime
    updated_at: datetime


class SkillProfile(BaseModel):
    id: str
    name: str
    category: SkillCategory
    level: CapabilityLevel = CapabilityLevel.beginner
    confidence: ConfidenceLevel = ConfidenceLevel.medium
    evidence: List[str] = []
    related_projects: List[str] = []
    last_practiced_at: Optional[date] = None
    target_level: CapabilityLevel = CapabilityLevel.working
    gap_notes: str = ""
    priority: PriorityLevel = PriorityLevel.medium


class SkillStackSnapshotCreate(BaseModel):
    summary: str = ""
    skills: List[SkillProfile] = []


class SkillStackSnapshot(SkillStackSnapshotCreate):
    id: str
    created_at: datetime


class TaskProfile(BaseModel):
    id: str
    title: str
    category: TaskCategory
    status: TaskStatus = TaskStatus.planned
    current_phase: str = ""
    related_skill_ids: List[str] = []
    deadline: Optional[date] = None
    weekly_slot: str = ""
    progress_summary: str = ""
    blockers: List[str] = []
    next_milestone: str = ""
    evidence_outputs: List[str] = []
    adjustability: Adjustability = Adjustability.medium


class TaskStateSnapshotCreate(BaseModel):
    current_phase: str = ""
    summary: str = ""
    tasks: List[TaskProfile] = []


class TaskStateSnapshot(TaskStateSnapshotCreate):
    id: str
    created_at: datetime


class JDFitAnalysis(BaseModel):
    id: str
    jd_entry_id: str
    skill_snapshot_id: str
    task_snapshot_id: str
    role_type: JDRoleType = JDRoleType.unclear
    match_level: MatchLevel = MatchLevel.medium
    interest_adjusted_priority: PriorityLevel = PriorityLevel.medium
    timing_recommendation: TimingRecommendation = TimingRecommendation.pause_and_prepare
    technical_overlap: List[str] = []
    missing_skills: List[str] = []
    fatal_gaps: List[str] = []
    trainable_gaps: List[str] = []
    evidence_needed: List[str] = []
    red_flags: List[str] = []
    tpm_risk_level: RiskLevel = RiskLevel.unknown
    company_style_risk: RiskLevel = RiskLevel.unknown
    resume_suggestions: List[str] = []
    interview_selling_points: List[str] = []
    questions_to_ask_recruiter: List[str] = []
    recommendation: str = ""
    reasoning: str = ""
    created_at: datetime


class CandidateAction(BaseModel):
    id: str
    source_jd_ids: List[str] = []
    source_analysis_ids: List[str] = []
    related_skill_ids: List[str] = []
    related_task_ids: List[str] = []
    action_type: CandidateActionType
    title: str
    reason: str = ""
    expected_output: str = ""
    effort_estimate_min: int = 30
    priority: PriorityLevel = PriorityLevel.medium
    time_sensitivity: TimeSensitivity = TimeSensitivity.this_month
    risk: str = ""
    suggested_slot: str = ""
    status: CandidateActionStatus = CandidateActionStatus.suggested
    user_decision_reason: str = ""
    converted_task_id: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class CandidateActionDecision(BaseModel):
    reason: str = ""
    converted_task_id: Optional[str] = None


class JDFitAnalysisResult(BaseModel):
    analysis: JDFitAnalysis
    candidate_actions: List[CandidateAction] = []


class SkillKeywordTrend(BaseModel):
    keyword: str
    category: str = ""
    month_counts: Dict[str, int] = {}
    trend: str = "unknown"
    notes: str = ""


class JDDiscussionMessage(BaseModel):
    role: str
    content: str


class JDDiscussionRequest(BaseModel):
    messages: List[JDDiscussionMessage] = []
    content: str


class JDDiscussionResponse(BaseModel):
    content: str

File: ios/InfraAgent/Models/JDIntelligence.swift
import Foundation

struct JDEntry: Codable, Identifiable {
    let id: String
    let company: String
    let teamOrDepartment: String
    let roleTitle: String
    let city: String
    let sourceType: String
    let sourceName: String
    let sourceURL: URL?
    let recordDate: String
    let jdText: String
    let recruiterContext: String
    let notes: String
    let mustHaveSkills: [String]
    let bonusSkills: [String]
    let newKeywords: [String]
    let salaryRange: String
    let salaryMinK: Int?
    let salaryMaxK: Int?
    let salaryMonths: String
    let equity: String
    let educationRequirement: String
    let majorRequirement: String
    let experienceYears: String
    let paperOrPatentRequirement: String
    let domainExperienceRequirement: String
    let roleOrientation: String
    let acceptsTransition: String
    let languageOrOverseasRequirement: String
    let monthChange: String
    let personalGapAssessment: String
    let applicationPriority: String
    let status: String
    let createdAt: Date
    let updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case company
        case teamOrDepartment = "team_or_department"
        case roleTitle = "role_title"
        case city
        case sourceType = "source_type"
        case sourceName = "source_name"
        case sourceURL = "source_url"
        case recordDate = "record_date"
        case jdText = "jd_text"
        case recruiterContext = "recruiter_context"
        case notes
        case mustHaveSkills = "must_have_skills"
        case bonusSkills = "bonus_skills"
        case newKeywords = "new_keywords"
        case salaryRange = "salary_range"
        case salaryMinK = "salary_min_k"
        case salaryMaxK = "salary_max_k"
        case salaryMonths = "salary_months"
        case equity
        case educationRequirement = "education_requirement"
        case majorRequirement = "major_requirement"
        case experienceYears = "experience_years"
        case paperOrPatentRequirement = "paper_or_patent_requirement"
        case domainExperienceRequirement = "domain_experience_requirement"
        case roleOrientation = "role_orientation"
        case acceptsTransition = "accepts_transition"
        case languageOrOverseasRequirement = "language_or_overseas_requirement"
        case monthChange = "month_change"
        case personalGapAssessment = "personal_gap_assessment"
        case applicationPriority = "application_priority"
        case status
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }

    func withStatus(_ status: String) -> JDEntry {
        JDEntry(
            id: id,
            company: company,
            teamOrDepartment: teamOrDepartment,
            roleTitle: roleTitle,
            city: city,
            sourceType: sourceType,
            sourceName: sourceName,
            sourceURL: sourceURL,
            recordDate: recordDate,
            jdText: jdText,
            recruiterContext: recruiterContext,
            notes: notes,
            mustHaveSkills: mustHaveSkills,
            bonusSkills: bonusSkills,
            newKeywords: newKeywords,
            salaryRange: salaryRange,
            salaryMinK: salaryMinK,
            salaryMaxK: salaryMaxK,
            salaryMonths: salaryMonths,
            equity: equity,
            educationRequirement: educationRequirement,
            majorRequirement: majorRequirement,
            experienceYears: experienceYears,
            paperOrPatentRequirement: paperOrPatentRequirement,
            domainExperienceRequirement: domainExperienceRequirement,
            roleOrientation: roleOrientation,
            acceptsTransition: acceptsTransition,
            languageOrOverseasRequirement: languageOrOverseasRequirement,
            monthChange: monthChange,
            personalGapAssessment: personalGapAssessment,
            applicationPriority: applicationPriority,
            status: status,
            createdAt: createdAt,
            updatedAt: Date()
        )
    }
}

struct LocalJDPreferenceStore {
    private static let favoriteKey = "jd_local_favorite_ids"
    private static let favoriteRemovedKey = "jd_local_favorite_removed_ids"
    private static let interestedKey = "jd_local_interested_ids"
    private static let interestedRemovedKey = "jd_local_interested_removed_ids"

    static func isFavorite(_ entry: JDEntry) -> Bool {
        if favoriteRemovedIDs.contains(entry.id) {
            return false
        }
        return entry.status == "favorite" || favoriteIDs.contains(entry.id)
    }

    static func isFavorite(_ entryID: String) -> Bool {
        favoriteIDs.contains(entryID) && !favoriteRemovedIDs.contains(entryID)
    }

    static func isInterested(_ entry: JDEntry) -> Bool {
        if interestedRemovedIDs.contains(entry.id) {
            return false
        }
        return entry.status == "watching" || interestedIDs.contains(entry.id)
    }

    static func isInterested(_ entryID: String) -> Bool {
        interestedIDs.contains(entryID) && !interestedRemovedIDs.contains(entryID)
    }

    static func setFavorite(_ isFavorite: Bool, entryID: String) {
        var ids = favoriteIDs
        var removed = favoriteRemovedIDs
        if isFavorite {
            ids.insert(entryID)
            removed.remove(entryID)
        } else {
            ids.remove(entryID)
            removed.insert(entryID)
        }
        UserDefaults.standard.set(Array(ids), forKey: favoriteKey)
        UserDefaults.standard.set(Array(removed), forKey: favoriteRemovedKey)
    }

    static func setInterested(_ isInterested: Bool, entryID: String) {
        var ids = interestedIDs
        var removed = interestedRemovedIDs
        if isInterested {
            ids.insert(entryID)
            removed.remove(entryID)
        } else {
            ids.remove(entryID)
            removed.insert(entryID)
        }
        UserDefaults.standard.set(Array(ids), forKey: interestedKey)
        UserDefaults.standard.set(Array(removed), forKey: interestedRemovedKey)
    }

    static func apply(to entries: [JDEntry]) -> [JDEntry] {
        entries.map { entry in
            if isFavorite(entry) {
                return entry.withStatus("favorite")
            }
            if isInterested(entry) && entry.status != "favorite" {
                return entry.withStatus("watching")
            }
            if entry.status == "favorite" || entry.status == "watching" {
                return entry.withStatus("new")
            }
            return entry
        }
    }

    static func preferenceTags(for entry: JDEntry) -> [String] {
        var tags: [String] = []
        if isFavorite(entry) {
            tags.append("favorite")
        }
        if isInterested(entry) {
            tags.append("interested")
        }
        return tags
    }

    private static var favoriteIDs: Set<String> {
        Set(UserDefaults.standard.stringArray(forKey: favoriteKey) ?? [])
    }

    private static var favoriteRemovedIDs: Set<String> {
        Set(UserDefaults.standard.stringArray(forKey: favoriteRemovedKey) ?? [])
    }

    private static var interestedIDs: Set<String> {
        Set(UserDefaults.standard.stringArray(forKey: interestedKey) ?? [])
    }

    private static var interestedRemovedIDs: Set<String> {
        Set(UserDefaults.standard.stringArray(forKey: interestedRemovedKey) ?? [])
    }
}

struct JDEntryCreate: Codable {
    let company: String
    let teamOrDepartment: String
    let roleTitle: String
    let city: String
    let sourceType: String
    let sourceName: String
    let sourceURL: URL?
    let jdText: String
    let recruiterContext: String
    let notes: String
    let mustHaveSkills: [String]
    let bonusSkills: [String]
    let newKeywords: [String]
    let salaryRange: String
    let roleOrientation: String
    let applicationPriority: String

    enum CodingKeys: String, CodingKey {
        case company
        case teamOrDepartment = "team_or_department"
        case roleTitle = "role_title"
        case city
        case sourceType = "source_type"
        case sourceName = "source_name"
        case sourceURL = "source_url"
        case jdText = "jd_text"
        case recruiterContext = "recruiter_context"
        case notes
        case mustHaveSkills = "must_have_skills"
        case bonusSkills = "bonus_skills"
        case newKeywords = "new_keywords"
        case salaryRange = "salary_range"
        case roleOrientation = "role_orientation"
        case applicationPriority = "application_priority"
    }
}

struct JDImageImportRequest: Codable {
    let imageBase64: String
    let filename: String
    let sourceName: String

    enum CodingKeys: String, CodingKey {
        case imageBase64 = "image_base64"
        case filename
        case sourceName = "source_name"
    }
}

struct JDImageImportResult: Codable {
    let ocrText: String
    let draftEntry: JDEntryCreate
    let confidence: String
    let notes: String

    enum CodingKeys: String, CodingKey {
        case ocrText = "ocr_text"
        case draftEntry = "draft_entry"
        case confidence
        case notes
    }
}

struct JDPreferenceMarkCreate: Codable {
    let interestLevel: String
    let fitFeeling: String
    let userTags: [String]
    let whyLiked: String
    let whyHesitated: String
    let followUpIntent: String

    enum CodingKeys: String, CodingKey {
        case interestLevel = "interest_level"
        case fitFeeling = "fit_feeling"
        case userTags = "user_tags"
        case whyLiked = "why_liked"
        case whyHesitated = "why_hesitated"
        case followUpIntent = "follow_up_intent"
    }
}

struct JDPreferenceMark: Codable, Identifiable {
    let id: String
    let jdEntryID: String
    let interestLevel: String
    let fitFeeling: String
    let userTags: [String]
    let whyLiked: String
    let whyHesitated: String
    let followUpIntent: String
    let createdAt: Date
    let updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case jdEntryID = "jd_entry_id"
        case interestLevel = "interest_level"
        case fitFeeling = "fit_feeling"
        case userTags = "user_tags"
        case whyLiked = "why_liked"
        case whyHesitated = "why_hesitated"
        case followUpIntent = "follow_up_intent"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

struct SkillStackSnapshot: Codable, Identifiable {
    let id: String
    let createdAt: Date
    let summary: String
    let skills: [SkillProfile]

    enum CodingKeys: String, CodingKey {
        case id
        case createdAt = "created_at"
        case summary
        case skills
    }
}

struct SkillProfile: Codable, Identifiable {
    let id: String
    let name: String
    let category: String
    let level: String
    let confidence: String
    let evidence: [String]
    let relatedProjects: [String]
    let lastPracticedAt: String?
    let targetLevel: String
    let gapNotes: String
    let priority: String

    enum CodingKeys: String, CodingKey {
        case id
        case name
        case category
        case level
        case confidence
        case evidence
        case relatedProjects = "related_projects"
        case lastPracticedAt = "last_practiced_at"
        case targetLevel = "target_level"
        case gapNotes = "gap_notes"
        case priority
    }
}

struct TaskStateSnapshot: Codable, Identifiable {
    let id: String
    let createdAt: Date
    let currentPhase: String
    let summary: String
    let tasks: [TaskProfile]

    enum CodingKeys: String, CodingKey {
        case id
        case createdAt = "created_at"
        case currentPhase = "current_phase"
        case summary
        case tasks
    }
}

struct TaskProfile: Codable, Identifiable {
    let id: String
    let title: String
    let category: String
    let status: String
    let currentPhase: String
    let relatedSkillIDs: [String]
    let deadline: String?
    let weeklySlot: String
    let progressSummary: String
    let blockers: [String]
    let nextMilestone: String
    let evidenceOutputs: [String]
    let adjustability: String

    enum CodingKeys: String, CodingKey {
        case id
        case title
        case category
        case status
        case currentPhase = "current_phase"
        case relatedSkillIDs = "related_skill_ids"
        case deadline
        case weeklySlot = "weekly_slot"
        case progressSummary = "progress_summary"
        case blockers
        case nextMilestone = "next_milestone"
        case evidenceOutputs = "evidence_outputs"
        case adjustability
    }
}

struct JDFitAnalysis: Codable, Identifiable {
    let id: String
    let jdEntryID: String
    let skillSnapshotID: String
    let taskSnapshotID: String
    let roleType: String
    let matchLevel: String
    let interestAdjustedPriority: String
    let timingRecommendation: String
    let technicalOverlap: [String]
    let missingSkills: [String]
    let fatalGaps: [String]
    let trainableGaps: [String]
    let evidenceNeeded: [String]
    let redFlags: [String]
    let tpmRiskLevel: String
    let companyStyleRisk: String
    let resumeSuggestions: [String]
    let interviewSellingPoints: [String]
    let questionsToAskRecruiter: [String]
    let recommendation: String
    let reasoning: String
    let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case jdEntryID = "jd_entry_id"
        case skillSnapshotID = "skill_snapshot_id"
        case taskSnapshotID = "task_snapshot_id"
        case roleType = "role_type"
        case matchLevel = "match_level"
        case interestAdjustedPriority = "interest_adjusted_priority"
        case timingRecommendation = "timing_recommendation"
        case technicalOverlap = "technical_overlap"
        case missingSkills = "missing_skills"
        case fatalGaps = "fatal_gaps"
        case trainableGaps = "trainable_gaps"
        case evidenceNeeded = "evidence_needed"
        case redFlags = "red_flags"
        case tpmRiskLevel = "tpm_risk_level"
        case companyStyleRisk = "company_style_risk"
        case resumeSuggestions = "resume_suggestions"
        case interviewSellingPoints = "interview_selling_points"
        case questionsToAskRecruiter = "questions_to_ask_recruiter"
        case recommendation
        case reasoning
        case createdAt = "created_at"
    }
}

struct CandidateAction: Codable, Identifiable {
    let id: String
    let sourceJDIDs: [String]
    let sourceAnalysisIDs: [String]
    let relatedSkillIDs: [String]
    let relatedTaskIDs: [String]
    let actionType: String
    let title: String
    let reason: String
    let expectedOutput: String
    let effortEstimateMin: Int
    let priority: String
    let timeSensitivity: String
    let risk: String
    let suggestedSlot: String
    let status: String
    let userDecisionReason: String
    let convertedTaskID: String?
    let createdAt: Date
    let updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case sourceJDIDs = "source_jd_ids"
        case sourceAnalysisIDs = "source_analysis_ids"
        case relatedSkillIDs = "related_skill_ids"
        case relatedTaskIDs = "related_task_ids"
        case actionType = "action_type"
        case title
        case reason
        case expectedOutput = "expected_output"
        case effortEstimateMin = "effort_estimate_min"
        case priority
        case timeSensitivity = "time_sensitivity"
        case risk
        case suggestedSlot = "suggested_slot"
        case status
        case userDecisionReason = "user_decision_reason"
        case convertedTaskID = "converted_task_id"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }

    func withConvertedStatus(note: String) -> CandidateAction {
        withStatus("converted_to_task", note: note, convertedTaskID: "converted_\(id)")
    }

    func withStatus(_ status: String, note: String = "", convertedTaskID: String? = nil) -> CandidateAction {
        CandidateAction(
            id: id,
            sourceJDIDs: sourceJDIDs,
            sourceAnalysisIDs: sourceAnalysisIDs,
            relatedSkillIDs: relatedSkillIDs,
            relatedTaskIDs: relatedTaskIDs,
            actionType: actionType,
            title: title,
            reason: reason,
            expectedOutput: expectedOutput,
            effortEstimateMin: effortEstimateMin,
            priority: priority,
            timeSensitivity: timeSensitivity,
            risk: risk,
            suggestedSlot: suggestedSlot,
            status: status,
            userDecisionReason: note,
            convertedTaskID: convertedTaskID ?? self.convertedTaskID,
            createdAt: createdAt,
            updatedAt: Date()
        )
    }
}

struct LocalCandidateActionDecisionStore {
    private static let key = "jd_candidate_action_decisions"

    static func status(for actionID: String) -> String? {
        decisions[actionID]
    }

    static func save(_ action: CandidateAction) {
        var values = decisions
        values[action.id] = action.status
        UserDefaults.standard.set(values, forKey: key)
        if action.status == "converted_to_task" {
            LocalConvertedTaskStore.add(action, note: action.userDecisionReason)
        }
    }

    static func apply(to actions: [CandidateAction]) -> [CandidateAction] {
        actions.map { action in
            guard let status = status(for: action.id) else { return action }
            if status == "converted_to_task" {
                return action.withConvertedStatus(note: action.userDecisionReason)
            }
            return action.withStatus(status, note: action.userDecisionReason)
        }
    }

    private static var decisions: [String: String] {
        UserDefaults.standard.dictionary(forKey: key) as? [String: String] ?? [:]
    }
}

struct CandidateActionDecision: Codable {
    let reason: String
    let convertedTaskID: String?

    enum CodingKeys: String, CodingKey {
        case reason
        case convertedTaskID = "converted_task_id"
    }
}

struct JDFitAnalysisResult: Codable {
    let analysis: JDFitAnalysis
    let candidateActions: [CandidateAction]

    enum CodingKeys: String, CodingKey {
        case analysis
        case candidateActions = "candidate_actions"
    }
}

struct JDDiscussionMessagePayload: Codable {
    let role: String
    let content: String
}

struct JDDiscussionRequest: Codable {
    let messages: [JDDiscussionMessagePayload]
    let content: String
}

struct JDDiscussionResponse: Codable {
    let content: String
}
